tests/v4l2decode.cpp

```diff
@@ -330,6 +329,153 @@ static bool displayOneVideoFrameAndroid(int32_t fd, int32_t index)
+{
+    struct display * const d = &waylandDisplay;
+
+    if (!d->surface) {
+        d->surface.reset(wl_compositor_create_surface(d->compositor.get()), wl_surface_destroy);
+        if (!d->surface)
+            return false;
+    }
+
+    if (!d->shell_surface) {
+        d->shell_surface.reset(wl_shell_get_shell_surface(d->shell.get(), d->surface.get()),
+                                                                       wl_shell_surface_destroy);
+        if (!d->shell_surface)
+            return false;
+        wl_shell_surface_set_toplevel(d->shell_surface.get());
+    }
+    return true;
+}
+
+bool createWaylandDisplay()
+{
+    SharedPtr<VADisplay> display;
+    struct display *d = &waylandDisplay;
+    struct wl_registry_listener registry_listener = {
+        registryHandle,
+        NULL,
+    };
+
+    d->display.reset(wl_display_connect(NULL), wl_display_disconnect);
+    if (!d->display) {
+        return false;
+    }
+    wl_display_set_user_data(d->display.get(), d);
+    struct wl_registry *registry = wl_display_get_registry(d->display.get());
+    wl_registry_add_listener(registry, &registry_listener, d);
+    wl_display_dispatch(d->display.get());
+    return true;
+}
+
+bool vaPutSurfaceWayland(struct wl_buffer *buffer,
+                                              const VARectangle *srcRect,
+                                              const VARectangle *dstRect)
+{
+    struct wl_callback *callback;
+    struct display * const d = &waylandDisplay;
+    struct wl_callback_listener frame_callback_listener = {frameRedrawCallback};
+
+    if (redrawPending) {
+        wl_display_flush(d->display.get());
+        while (redrawPending) {
+            wl_display_dispatch(d->display.get());
+        }
+    }
+
+    if (!ensureWindow(dstRect->width, dstRect->height))
+        return false;
+
+    wl_surface_attach(d->surface.get(), buffer, 0, 0);
+    wl_surface_damage(d->surface.get(), dstRect->x,
+		dstRect->y, dstRect->width, dstRect->height);
+    wl_display_flush(d->display.get());
+    redrawPending = true;
+    callback = wl_surface_frame(d->surface.get());
+    wl_callback_add_listener(callback, &frame_callback_listener,&redrawPending);
+    wl_surface_commit(d->surface.get());
+    return true;
+}
+
+bool output(struct wl_buffer *buffer)
+{
+    VARectangle srcRect, dstRect;
+    if (!ensureWindow(videoWidth, videoHeight))
+        return false;
+
+    srcRect.x = 0;
+    srcRect.y = 0;
+    srcRect.width  = videoWidth;
+    srcRect.height = videoHeight;
+
+    dstRect.x = 0;
+    dstRect.y = 0;
+    dstRect.width  = videoWidth;
+    dstRect.height = videoHeight;
+    return vaPutSurfaceWayland(buffer, &srcRect, &dstRect);
+}
+
+static bool displayOneVideoFrameWayland(int32_t fd, struct v4l2_buffer  *buf)
+{
+    int32_t ioctlRet = -1;
+    struct v4l2_buffer buffer;
+
+    output((struct wl_buffer *)(buf->m.userptr));
+    memset(&buffer, 0, sizeof(buffer));
```

Comment:
in fact, takeOneOutputFrame() does VIDIOC_QBUF.
we needn't QBUF here, nor for displayOneVideoFrameAndroid. (just be careful that buffer index may change).